stew/io2.nim

```diff
@@ -0,0 +1,1003 @@
+  proc getCurrentDirectoryW(nBufferLength: uint32,
+                            lpBuffer: WideCString): uint32 {.
+       importc: "GetCurrentDirectoryW", dynlib: "kernel32", stdcall,
+       sideEffect.}
+  proc formatMessageW(dwFlags: uint32, lpSource: pointer,
+                      dwMessageId, dwLanguageId: uint32,
+                      lpBuffer: pointer, nSize: uint32,
+                      arguments: pointer): uint32 {.
+       importc: "FormatMessageW", stdcall, dynlib: "kernel32".}
+  proc localFree(p: pointer): uint {.
+       importc: "LocalFree", stdcall, dynlib: "kernel32".}
+  # proc getTempPathW(nBufferLength: uint32, lpBuffer: WideCString): uint32 {.
+  #      importc: "GetTempPathW", dynlib: "kernel32", stdcall.}
+  # proc getUserProfileDirectoryW(hToken: uint, lpProfileDir: WideCString,
+  #                               lpcchSize: var uint32): uint32 {.
+  #      importc: "GetUserProfileDirectoryW", dynlib: "userenv.dll", stdcall.}
+  proc getLongPathNameW(lpszShortPath: WideCString, lpszLongPath: WideCString,
+                        cchBuffer: uint32): uint32 {.
+       importc: "GetLongPathNameW", dynlib: "kernel32.dll", stdcall.}
+  # proc getCurrentProcessToken(): uint =
+  #   # (HANDLE)(LONG_PTR) -4;
+  #   cast[uint](-4)
+
+  proc `==`*(a: IoErrorCode, b: uint32): bool {.inline.} =
+    (uint32(a) == b)
+
+elif defined(posix):
+  import posix
+
+  const
+    DirSep* = '/'
+    AltSep* = '/'
+    BothSeps* = {'/'}
+
+  type
+    IoHandle* = distinct cint
+    IoErrorCode* = distinct cint
+
+  when defined(linux):
+    const
+      O_DIRECT = cint(0x4000)
+      O_CLOEXEC = cint(0x2000000)
+  elif defined(freebsd):
+    const
+      O_DIRECT = cint(0x10000)
+      O_CLOEXEC = cint(0x100000)
+  elif defined(dragonflybsd):
+    const
+      O_DIRECT = cint(0x10000)
+      O_CLOEXEC = cint(0x20000)
+  elif defined(netbsd):
+    const
+      O_DIRECT = cint(0x80000)
+      O_CLOEXEC = cint(0x400000)
+  elif defined(openbsd):
+    const
+      O_CLOEXEC = cint(0x10000)
+  elif defined(macosx):
+    const
+      O_CLOEXEC = cint(0x1000000)
+      F_NOCACHE = cint(48)
+
+  var errno {.importc, header: "<errno.h>".}: cint
+
+  proc write(a1: cint, a2: pointer, a3: csize_t): int {.
+       importc, header: "<unistd.h>".}
+  proc read(a1: cint, a2: pointer, a3: csize_t): int {.
+       importc, header: "<unistd.h>".}
+  proc c_strlen(a: cstring): cint {.
+       importc: "strlen", header: "<string.h>", noSideEffect.}
+  proc c_strerror(errnum: cint): cstring {.
+       importc: "strerror", header: "<string.h>".}
+  proc getcwd(a1: cstring, a2: int): cstring {.
+       importc, header: "<unistd.h>", sideEffect.}
+  proc `==`*(a: IoErrorCode, b: cint): bool {.inline.} =
+    (cint(a) == b)
+
+type
+  IoResult*[T] = Result[T, IoErrorCode]
+
+  OpenFlags* = enum
+    ReadOnly, WriteOnly, ReadWrite, Create, Exclusive, Append, Truncate,
+    NoInherit, NonBlock, Direct
```

Comment:
I think it'd be better if we use `Read` and `Write` flags instead of having three different flags.

Also for safety reasons I'd prefer having `NoInherit` being the default. The stdlib has already make this change.

Reply:
@alaviss from my point of view `stdlib` is incorrect because of this, it should not make any decisions on what should be default and what should not. It should allow developers to choose their own way.

Reply:
It's the wrong default. The reason why most operating system API doesn't use `NoInherit` by default is purely historical. They can't change that now due to compatibility reasons, but as a (relatively new) `stdlib` we can, and we did.

Nowadays people have to add syscalls like `closefrom`, `close_range` just to walkaround the historical baggage of not having `NoInherit` (`O_CLOEXEC`) as the default. Even Microsoft has been making all "new" APIs in Windows have inheritance disabled by default (`bInheritHandles` defaults to 0 when not provided for many APIs). And I doubt any of them were opinionated, it's just the reality that FD inheritance was the wrong default, and one that bites.